core/src/cloud/grow/buffered-event-stream.ts

```diff
@@ -0,0 +1,319 @@
+    )
+
+    this.eventBuffer.set(event.eventId, event)
+    void this.outputStream?.write(event)
+  }
+
+  private handleCommandCompleted(event: GrpcEvent) {
+    event.eventData = {
+      case: "commandEvent",
+      value: create(GardenCommandEventSchema, {
+        // TODO: use ulid
+        // TODO: add sessionId to the event payload, otherwise this will not behave as expected
+        commandId: this.garden.sessionId,
+
+        // completed now
+        completedAt: timestampFromDate(new Date()),
+        status: GardenCommandEvent_Status.SUCCEEDED,
+      }),
+    }
+  }
+
+  private handleCommandStarted(event: GrpcEvent, payload: CommandInfoPayload) {
+    event.eventData = {
+      case: "commandEvent",
+      value: create(GardenCommandEventSchema, {
+        // TODO: use ulid
+        commandId: payload.sessionId,
+
+        // started now
+        startedAt: timestampFromDate(new Date()),
+        status: GardenCommandEvent_Status.IN_PROGRESS,
+
+        // TODO: expose if it is a custom command
+        //isCustomCommand: payload.isCustomCommand,
+
+        invocationMetadata: create(GardenCommandEvent_InvocationMetadataSchema, {
+          cwd: process.cwd(),
+          command: payload.name,
+          // TODO: args
+          //args: payload.args,
+          // TODO: tty detection
+          //interactiveTty: payload.interactiveTty,
+        }),
+
+        gitMetadata: create(GardenCommandEvent_GitMetadataSchema, {
+          // TODO: repository root
+          //repositoryRoot: payload.repositoryRoot,
+          headRefSha: payload.vcsCommitHash,
+          // TODO: expose raw ref name, might be a tag
+          // TODO: might be optional if the head is detached
+          headRefName: `refs/heads/${payload.vcsBranch}`,
+          // TODO: expose all remotes with their original names
+          gitRemotes: [
+            create(GardenCommandEvent_GitMetadata_GitRemoteSchema, {
+              name: "origin",
+              url: payload.vcsOriginUrl,
+            }),
+          ],
+        }),
+
+        gardenProjectMetadata: create(GardenCommandEvent_GardenProjectMetadataSchema, {
+          projectName: payload.projectName,
+
+          // TODO: expose all project data
+
+          // projectApiVersion: payload.projectApiVersion,
+
+          // projectId: payload.projectId,
+          // projectRootDir: payload.projectRootDir,
+
+          // namespaceName: payload.namespaceName,
+          // namespaceId: payload.namespaceId,
+
+          // environmentName: payload.environmentName,
+          // environmentId: payload.environmentId,
+        }),
+      }),
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  private handleLogEntry(logEntry: LogEntryEventPayload) {
+    // TODO: logs handling
+  }
+
+  private async streamEvents() {
+    this.outputStream = createWritableIterable<GrpcEvent>()
+
+    const ackStream = this.client.ingestEventStream(this.outputStream)
```

Comment:
Instead of calling this "client" it might be a more useful mental model if we imagine just calling the method on the service directly.
```suggestion
    const ackStream = this.eventService.ingestEventStream(this.outputStream)
```